### PR TITLE
add announcement generator

### PR DESCRIPTION
**Pull request by MarcelHMendes**

### CODE REVIEW

File: objgen.py
Comment (cunha) [inline objgen.py]: I know this is OK for `dataclasses`, but not for normal classes. Just double check this is Pythonic.

File: objgen.py
Comment (cunha) [inline objgen.py]: Add annotation for return type?

File: objgen.py
Comment (cunha) [inline objgen.py]: It seems like this whole function can be turned into a one-liner by using `**kwargs`.

File: objgen.py
Comment (cunha) [inline objgen.py]: I am not sure I follow what this class does.  I think what we would want is one class for each entity people can reason about, and then create methods in each class to generate instances of that class.  Classes that make sense to me:

* PrefixAnnouncement:  Specifies how one prefix should be announced.  Contains parameters such as the origin ASN, the number of prepends, and the upstream muxes.  We could also consider adding poisoning as a parameter, even if it will not be allowed in the REST API.

* ExperimentRound:  Just a `list[PrefixAnnouncement]`

* Experiment:  Just a `list[ExperimentRound]`

Comment (cunha) [reply objgen.py]: I think we could create `dataclasses` for each of these, as they mostly hold information, and then provide verification- and initialization-related functions in the `dataclass`.

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json:78]: We should discuss whether we want to include BGP poisoning here, and disallow it in the API.

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json]: Double-check these

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json]: I think we should add some predefined optional properties:

* description
* contact

And possibly other general-purpose fields.

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json:1]: There is a way in Git to rename a file. So you can apply changes to the existing `announcement_schema.json`, then rename the file, so that the changes will show up properly in the diff (i.e., it may be better to evaluate this as a diff on top of the existing file than a new one).

File: objgen.py
Comment (MarcelHMendes) [reply objgen.py]: I'm wondering if we need to create a data class for the Experiment Round, since a round is a list of announcements. My idea was to create a method that adds a round passing a list of announcements as arguments.

File: requirements.txt
Comment (cunha) [inline requirements.txt]: I think `pexpect` and `beautifulsoup` are needed only for the scripts in `utils`, so maybe we should move these two dependencies in another `requirements.txt` file into that directory.  This way people can install the minimum necessary for the tools they will use.

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json:5]: We should think about how to allow additional properties.  In particular, we want people to add descriptions, their contact addresses, and other useful information to their experiments.  Maybe we should make `description` and `email` mandatory fields in an experiment.

File: configs/announcement_schema.json
Comment (cunha) [inline configs/announcement_schema.json]: For `prefix` above, I think some recent version of JSON Schema have defined an IP address class.  We could look into using that instead of rolling our own pattern checks.  If we go for the pattern checks, then we should update to allow IPv6 too.  IPv6 is more important than IPv4 these days.

File: peering.py
Comment (cunha) [inline peering.py]: OK, but this file is missing some `argparse` and the command line args to call into `ExperimentController`.

File: peeringmod.py
Comment (cunha) [inline peeringmod.py]: When overwriting `__init__`, we should not use `dataclasses`.  The main idea behind `dataclasses` is that it will do the initialization for us (here we're doing initialization by hand with `__init__` and `add_round`).  I think we want to either remove `__init__` and `add_round` *or* remove `dataclass`; having both does not make sense.

File: peeringmod.py
Comment (cunha) [inline peeringmod.py]:
```{python}
dictdetails = dict(kwargs)
dictdetails["muxes"] = muxes
self.details.append(dictdetails)
```

But even then I am not a fan of this design.  It seems like all of the logic is bundled in the `details` field.  I'd prefer if we made things more explicit.  I mean, how are the different details used?

File: peeringmod.py
Comment (cunha) [inline peeringmod.py]: Alright, so it seems we're keeping the announcement around as a dictionary.  Should we instead create something like

```
@dataclasses.dataclass
class PrefixAnnouncement:
    prefix: IPv4Network | IPv6Network
    withdrawals: list[Mux]
    announcements: list[Announcement]

    def validate(schema_fn: pathlib.Path) -> bool:
        ...
```

Comment (cunha) [reply peeringmod.py]: Additionally, we could also look into avoiding the need for `validate` altogether.  This combines two ideas:

1. Only allow the creation of valid objects by carefully checking any created object is valid *on construction*.  For example, we could declare an `enum Mux` and initialize the `enum` dynamically at runtime by reading the valid muxes from the JSON Schema (see [this link][so-dynamic-enum]).  This way people can only initialize instances of `Mux` that are valid.

[so-dynamic-enum]: https://stackoverflow.com/questions/43096541/a-more-pythonic-way-to-define-an-enum-with-dynamic-members

2. Validate instances read from files when reading (we can load the JSON, validate, and *then* initialize the classes from the JSON).  The idea is that users don't need to explicitly call `validate`.

**Review by cunha — CHANGES_REQUESTED** [review]

File: configs/announcement_schema.json
Comment (cunha) [reply configs/announcement_schema.json]: It also seems this file does not have feature-parity with the file in the site-dockerized repo. I think we should make the two files the same, if possible. This way we would need to deal with a single type of JSON. If it is not possible to use the same schema, we should document the differences. One particular thing I am missing in this version is

```{json}
      "propertyNames": {
        "$ref": "#/definitions/allocatedPrefix"
      },
```

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json]: We can avoid the CamelCase by setting the name of the property to just `description`.

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json]: Can we remove the `maxLength`? Otherwise let's bump it to 4096.

File: utils/experiment-examples/experiment-1.json
Comment (cunha) [inline utils/experiment-examples/experiment-1.json:5]: Add `email` and `description`. In the description, provide an example of what will happen. Overall, make the example realistic (this one seems to make the exact same announcement 4 times, for no apparent reason).

One possible example would be to use two prefixes to test propagation delay of an announcement from `ufmg01`, `ams01`, `seattle01` and anycast. This would require 3 rounds and 2 prefixes.

File: peering.py
Comment (cunha) [inline peering.py]: Should this be a parameter to `__init__`?

File: peering.py
Comment (cunha) [inline peering.py]: Please provide some documentation. Is the access token downloaded from the website's dashboard? What is `refresh`?

Comment (cunha) [reply peering.py]: It seems we can provide a default for `schema_fn`, as there is basically one file users will want to use as the schema.

Comment (cunha) [reply peering.py]: We should also pass a default to `schema_fn` in the `AnnouncementController` above.

**Review by cunha — CHANGES_REQUESTED** [review]
Left some suggestions inline

File: peering.py
Comment (MarcelHMendes) [reply peering.py]: > Is the access token downloaded from the website's dashboard?
We can get it via browser:
image
or api request.

Comment (cunha) [reply peering.py]: I don't recognize that screenshot. Is it from the Website? The GET request seems to have failed, so not sure what it's intending to show. What I meant with my question is that users currently check their `AnnouncementCapabilities`, prefixes, deployed containers, and download their certs from the dashboard. So I thought we might make the token available right there... although we could discuss alternatives.

Comment (MarcelHMendes) [reply peering.py]: This page is generated by django, the message above is just saying that the get request is not accepted, which is correct because to obtain the token you need to make a post request with username and password.

Comment (MarcelHMendes) [reply peering.py]: With django-restframerok we also generate endpoints for users to send their experiments via web browser  (if they're already logged in ) and obtain tokens from web browsers too.

**Review by cunha — APPROVED** [review]

**Comment by cunha** [comment]
Was going to merge this, but can't because it's marked as a draft. @MarcelHMendes are you planning on pushing more commits?

**Comment by MarcelHMendes** [comment]
I believe I will not change this code anymore, but I am not sure. Let's merge this together with the client-container's and integration-test's PR.

File: deploy.py
Comment (cunha) [inline deploy.py]: I don't think this works. `sys.argv` may have `--experiment=path` instead, and that is completely valid. It seems we may want to create subcommands, and have `--url` be a subcommand of `--experiment`.

**Review by cunha — COMMENTED** [review]
Looks good, but I don't think the `required` on `--url` will work as intended.

File: scripts/peering-bgp
Comment (cunha) [inline scripts/peering-bgp]: Please submit this change in a separate PR in a different branch so we can merge it immediately. (This PR is on hold until we merge and deploy the API to the website.)

**Review by cunha — APPROVED** [review]
PR looks good, but change to socket check should be in a different PR.

File: deploy.py
Comment (cunha) [reply deploy.py]: subcommands seem to be what we want to use: https://docs.python.org/3/library/argparse.html#sub-commands

File: utils/experiment-generator.py
Comment (cunha) [inline utils/experiment-generator.py]: Let's move this inside `utils/experiment-examples/` so it remains beside the file it generates.

File: deploy.py
Comment (cunha) [inline deploy.py]: Let us refactor this file inside `peering.py` so we have only one additional file in the top-level directory of the PEERING client.

File: configs/announcement_schema.json
Comment (cunha) [inline configs/announcement_schema.json]: I forget why we decided to add this; right now I think we can just drop this comment as experimenters are unlikely to edit this file.

File: configs/announcement_schema.json
Comment (cunha) [inline configs/announcement_schema.json:71]: Add: `saopaulo01` and `sbu01`
Remove: `phoenix01`

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json:4]: This seems to allow an experiment to have just a single `description` without any `rounds`. `rounds` is required. It seems we may need this: https://json-schema.org/understanding-json-schema/reference/object.html#required-properties

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json]: An item in `rounds` is not a `prefixAnnouncement`, it's an object mapping a prefix or letter to a `prefixAnnouncement`. The call to `jsonschema.validate()` in `peering.py` succeeds because it doesn't seem to be doing anything. You can pass anything to it and it works (see example below). I think we may be better off just having a single file with all the schemas inside. Then we can refer to them with the `"$ref": "#/path/to/element"` syntax.

```
>>> eg = peering.ExperimentController("x", "y")
>>> eg.schema
{'title': 'Experiment schema', 'type': 'object', 'minProperties': 1, 'additionalProperties': False, 'properties': {'rounds': {'type': 'array', 'minItems': 1, 'items': {'url': './announcement_schema.json#/definitions/prefixAnnouncement'}}, 'description': {'type': 'string'}, 'email': {'type': 'string', 'format': 'email'}}}
>>> exp
{'experiment': {'description': 'Testing prefix propagation of muxes amsterdam01, ufmg01 and seattle01', 'email': 'noc@peering.ee.columbia.edu', 'rounds': [{'wtf': 1}]}}
>>> jsonschema.validate(exp["experiment"], eg.schema)
>>>
```

**Review by cunha — CHANGES_REQUESTED** [review]

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json:30]: Let's add `ˆ` and `$` on the `[ABCD]` branch too?

File: configs/experiment_schema.json
Comment (cunha) [inline configs/experiment_schema.json:39]: What's the reasoning behind these prefixes? Are these the prefixes we will allocate to the API?

**Review by cunha — APPROVED** [review]
Let's merge https://github.com/MarcelHMendes/client/pull/1 and then this one

**Review by cunha — APPROVED** [review]

**Pull request merged**
